lib/rage/controller/api.rb

```diff
@@ -283,4 +288,42 @@ def headers
     @__headers = {}.merge!(@__headers) if DEFAULT_HEADERS.equal?(@__headers)
     @__headers
   end
+
+  class Request
+    SPECIAL_HEADERS = %w[CONTENT_TYPE CONTENT_LENGTH].freeze
+    # Get the request headers.
+    # @example
+    #  request.headers["Content-Type"] # => "application/json"
+    # or request.headers["HTTP_CONTENT_TYPE"] # => "application/json"
+    attr_reader :headers
+
+    def initialize(env)
+      @env = env
+    end
+
+    def headers
+      @headers ||= Headers.new(@env)
+    end
+
+    class Headers
+      def initialize(env)
+        @env = env
+        @headers = {}
+      end
+
+      def [](requested_header)
+        requested_header = requested_header.upcase.tr('-', '_')
```

Comment:
so I've run this benchmark:

```ruby
require "benchmark/ips"

requested_header = "Content-Type".freeze

Benchmark.ips do |x|
  x.report("1") { requested_header.upcase.tr!("-", "_") }
  x.report("2") { s = requested_header.tr("-", "_"); s.upcase! || s }

  x.compare!
end && true
```

and received the following result:

```
Warming up --------------------------------------
                   1   308.241k i/100ms
                   2   442.991k i/100ms
Calculating -------------------------------------
                   1      3.043M (± 1.5%) i/s -     15.412M in   5.066559s
                   2      4.575M (± 0.5%) i/s -     23.036M in   5.035321s

Comparison:
                   2:  4574912.3 i/s
                   1:  3042613.1 i/s - 1.50x  slower
```

Based on this, let's move this line inside the `else` branch and change it accordingly.